blockchain/message_pool/src/msgpool.rs

```diff
@@ -0,0 +1,1066 @@
+                }
+                base_sequence += 1;
+            }
+        }
+
+        Ok(base_sequence)
+    }
+
+    /// Get the state balance for the actor that corresponds to the supplied address and tipset,
+    /// if this actor does not exist, return an error
+    async fn get_state_balance(&mut self, addr: &Address, ts: &Tipset) -> Result<BigInt, Error> {
+        let api = self.api.read().await;
+        let actor = api.state_get_actor(&addr, &ts)?;
+        Ok(BigInt::from(actor.balance))
+    }
+
+    /// Remove a message given a sequence and address from the messagepool
+    pub async fn remove(&mut self, from: &Address, sequence: u64) -> Result<(), Error> {
+        remove(from, self.pending.clone(), sequence).await
+    }
+
+    /// Return a tuple that contains a vector of all signed messages and the current tipset for
+    /// self.
+    pub async fn pending(&self) -> Result<(Vec<SignedMessage>, Tipset), Error> {
+        let mut out: Vec<SignedMessage> = Vec::new();
+
+        let pending = self.pending.read().await;
+        let pending_hm = pending.clone();
+        drop(pending);
+
+        for (addr, _) in pending_hm {
+            out.append(
+                self.pending_for(&addr)
+                    .await
+                    .ok_or_else(|| Error::InvalidFromAddr)?
+                    .as_mut(),
+            )
+        }
+
+        let cur_t = self.cur_tipset.read().await;
+        let cur_ts = cur_t.clone();
+        drop(cur_t);
+
+        Ok((out, cur_ts))
+    }
+
+    /// Return a Vector of signed messages for a given from address. This vector will be sorted by
+    /// each messsage's sequence. If no corresponding messages found, return None result type
+    async fn pending_for(&self, a: &Address) -> Option<Vec<SignedMessage>> {
+        let pending = self.pending.read().await;
+
+        let mset = pending.get(a);
+        match mset {
+            Some(msgset) => {
+                if msgset.msgs.is_empty() {
+                    return None;
+                }
+
+                let mut msg_vec = Vec::new();
+
+                for (_, item) in msgset.msgs.clone() {
+                    msg_vec.push(item);
+                }
+
+                msg_vec.sort_by_key(|value| value.message().sequence());
+
+                Some(msg_vec)
+            }
+            None => None,
+        }
+    }
+
+    /// Return Vector of signed messages given a block header for self
+    pub async fn messages_for_blocks(
+        &mut self,
+        blks: &[BlockHeader],
+    ) -> Result<Vec<SignedMessage>, Error> {
+        let mut msg_vec: Vec<SignedMessage> = Vec::new();
+
+        for block in blks {
+            let api = self.api.read().await;
+            let (umsg, mut smsgs) = api.messages_for_block(&block)?;
+            drop(api);
+
+            msg_vec.append(smsgs.as_mut());
+            for msg in umsg {
+                let smsg = recover_sig(self.bls_sig_cache.clone(), msg).await?;
```

Comment:
if you change other point, you won't need to clone the Arc here, you would just get a read into the rwlock and pass the ref in